rest-client/client-vertx-impl/src/main/java/io/apicurio/registry/rest/client/VertxHttpClient.java

```diff
@@ -0,0 +1,211 @@
+    private final String basePath;
+    private final Vertx vertx;
+
+    private static final Map<String, String> DEFAULT_HEADERS = new HashMap<>();
+    private static final ThreadLocal<Map<String, String>> requestHeaders = ThreadLocal.withInitial(Collections::emptyMap);
+
+    public VertxHttpClient(String basePath, Map<String, Object> options, Auth auth) {
+        if (!basePath.endsWith("/")) {
+            basePath += "/";
+        }
+        this.vertx = Vertx.currentContext() != null ? Vertx.currentContext().owner() : Vertx.vertx();
+        this.webClient = WebClient.create(vertx);
+        this.auth = auth;
+        this.basePath = basePath;
+        addHeaders(options);
+    }
+
+    private static void addHeaders(Map<String, Object> configs) {
+
+        Map<String, String> requestHeaders = configs.entrySet().stream()
+                .filter(map -> map.getKey().startsWith(ClientConfig.REGISTRY_REQUEST_HEADERS_PREFIX))
+                .collect(Collectors.toMap(map -> map.getKey()
+                        .replace(ClientConfig.REGISTRY_REQUEST_HEADERS_PREFIX, ""), map -> map.getValue().toString()));
+
+        if (!requestHeaders.isEmpty()) {
+            requestHeaders.forEach(DEFAULT_HEADERS::put);
+        }
+    }
+
+    @Override
+    public <T> T sendRequest(Request<T> request) {
+        try {
+            final URI uri = buildURI(basePath + request.getRequestPath(), request.getPathParams());
+            final RequestOptions requestOptions = new RequestOptions();
+
+            requestOptions.setHost(uri.getHost());
+            requestOptions.setURI(uri.getPath());
+            requestOptions.setPort(uri.getPort());
+
+            DEFAULT_HEADERS.forEach(requestOptions::addHeader);
+
+            //Add current request headers
+            requestHeaders.get().forEach(requestOptions::addHeader);
+            requestHeaders.remove();
+
+            Map<String, String> headers = request.getHeaders();
+            if (this.auth != null) {
+                //make headers mutable...
+                headers = new HashMap<>(headers);
+                this.auth.apply(headers);
+            }
+            headers.forEach(requestOptions::addHeader);
+
+            CompletableFuture<T> resultHolder;
+
+            switch (request.getOperation()) {
+                case GET:
+                    resultHolder = executeGet(request, requestOptions);
+                    break;
+                case PUT:
+                    resultHolder = executePut(request, requestOptions);
+                    break;
+                case POST:
+                    resultHolder = executePost(request, requestOptions);
+                    break;
+                case DELETE:
+                    resultHolder = executeDelete(request, requestOptions);
+                    break;
+                default:
+                    throw new IllegalStateException("Operation not allowed");
+            }
+
+            return resultHolder.get();
```

Comment:
This is a blocking operation. Big fat _no_ on an event loop.

Reply:
Oh my, I forgot to use our util class.

Reply:
I'm afraid that has the same problem.

Reply:
Yes, you're right, just mixed some things.